src/main/java/seedu/address/logic/parser/FindCommandParser.java

```diff
@@ -7,57 +7,63 @@
 import static seedu.address.logic.parser.CliSyntax.PREFIX_PHONE;
 import static seedu.address.logic.parser.CliSyntax.PREFIX_TAG;
 
+import java.util.ArrayList;
 import java.util.List;
-import java.util.regex.Pattern;
+import java.util.function.Predicate;
 
-import seedu.address.logic.commands.AbstractFindCommand;
-import seedu.address.logic.commands.FindByEmailCommand;
-import seedu.address.logic.commands.FindByNameCommand;
-import seedu.address.logic.commands.FindByPhoneCommand;
-import seedu.address.logic.commands.FindByTagCommand;
+import seedu.address.logic.commands.SuperFindCommand;
 import seedu.address.logic.parser.exceptions.ParseException;
+import seedu.address.model.person.CombinedContainsKeywordsPredicate;
+import seedu.address.model.person.ContainsKeywordsPredicate;
 import seedu.address.model.person.EmailContainsKeywordsPredicate;
 import seedu.address.model.person.NameContainsKeywordsPredicate;
+import seedu.address.model.person.Person;
 import seedu.address.model.person.PhoneContainsKeywordsPredicate;
 import seedu.address.model.person.TagContainsKeywordsPredicate;
 
 /**
  * Parses input arguments and creates a new FindCommand object
  */
-public class FindCommandParser implements Parser<AbstractFindCommand> {
-
-    public static final Pattern KEYWORD_EXTRACTOR =
-            Pattern.compile("^(?<type>[pent]/)\\s*(?<arguments>[\\S\\s]+)$");
-
+public class FindCommandParser implements Parser<SuperFindCommand> {
     /**
      * Parses the given {@code String} of arguments in the context of the FindCommand
      * and returns a FindCommand object for execution.
      * @throws ParseException if the user input does not conform the expected format
      */
-    public AbstractFindCommand parse(String args) throws ParseException {
-        List<String> tempList;
-
+    @Override
+    public SuperFindCommand parse(String args) throws ParseException {
         requireNonNull(args);
-        ArgumentMultimap argMultimap =
-                ArgumentTokenizer.tokenize(args, PREFIX_NAME, PREFIX_PHONE, PREFIX_EMAIL, PREFIX_TAG);
+        ArgumentMultimap argMultimap = ArgumentTokenizer
+                .tokenize(args, PREFIX_NAME, PREFIX_PHONE, PREFIX_EMAIL, PREFIX_TAG);
+
+        List<Predicate<Person>> predicates = new ArrayList<>();
 
         if (argMultimap.getValue(PREFIX_NAME).isPresent()) {
-            tempList = argMultimap.getAllValues(PREFIX_NAME);
-            return new FindByNameCommand(new NameContainsKeywordsPredicate(tempList));
+            List<String> nameKeywords = argMultimap.getAllValues(PREFIX_NAME);
```

Comment:
Perhaps we could see if we can extract these prefix checks into their own methods to separate the level of abstraction.